torch/utils/_pytree.py

```diff
@@ -661,7 +661,7 @@ def _is_leaf(tree: PyTree, is_leaf: Optional[Callable[[PyTree], bool]] = None) -
 # context: some context that is useful in unflattening the pytree
 # children_specs: specs for each child of the root Node
 # num_leaves: the number of leaves
-@dataclasses.dataclass
+@dataclasses.dataclass(init=True, frozen=True, eq=True, repr=False)
```

Comment:
I do not enable `slots=True`.

1. `slots` argument only supports Python 3.11+.
2. The `num_{nodes,num_leaves}` attributes become descriptors during dynamo frame convertion if set `slots=True`. Then it will raise unimplemented error for `member_descriptor.__radd__(int)` in:

```python
num_nodes = sum((spec.num_nodes for spec in self.children_specs), start=1)
num_leaves = sum(spec.num_leaves for spec in self.children_specs)
```

Reply:
`init` and `eq` are set to their default while `repr` is not. We define a custom `__repr__` method below. I can remove `init` and `eq` if requested.

PS: I think we need to somehow resolve `__eq__` for `dict`-specs.

- #114392

Reply:
I'm fine with either, it's up to you